src/messages/textDocument_document.cc

```diff
@@ -165,88 +152,70 @@ void MessageHandler::textDocument_documentSymbol(JsonReader &reader,
+              rhs.extent.end < lhs.extent.end);
+    };
+
     for (auto [sym, refcnt] : file->symbol2refcnt) {
       if (refcnt <= 0 || !sym.extent.valid())
         continue;
-      auto r = sym2ds.try_emplace(SymbolIdx{sym.usr, sym.kind});
-      auto &ds = r.first->second;
-      if (!ds || sym.role & Role::Definition) {
-        if (!ds)
-          ds = std::make_unique<DocumentSymbol>();
-        if (auto range = getLsRange(wf, sym.range)) {
-          ds->selectionRange = *range;
-          ds->range = ds->selectionRange;
-          // For a macro expansion, M(name), we may use `M` for extent and
-          // `name` for spell, do the check as selectionRange must be a subrange
-          // of range.
-          if (sym.extent.valid())
-            if (auto range1 = getLsRange(wf, sym.extent);
-                range1 && range1->includes(*range))
-              ds->range = *range1;
-        }
+      syms.push_back(sym);
+    }
+    std::sort(std::begin(syms), std::end(syms), sym_cmp);
+
+    std::vector<DocumentSymbol> result;
+    std::stack<DocumentSymbol *> indent;
+    for (auto sym : syms) {
+      DocumentSymbol ds;
+      if (auto range = getLsRange(wf, sym.range)) {
+        ds.selectionRange = *range;
+        ds.range = ds.selectionRange;
+        // For a macro expansion, M(name), we may use `M` for extent and
+        // `name` for spell, do the check as selectionRange must be a subrange
+        // of range.
+        if (sym.extent.valid())
+          if (auto range1 = getLsRange(wf, sym.extent);
+              range1 && range1->includes(*range))
+            ds.range = *range1;
       }
-      if (!r.second)
-        continue;
-      std::vector<const void *> def_ptrs;
-      SymbolKind kind = SymbolKind::Unknown;
       withEntity(db, sym, [&](const auto &entity) {
-        auto *def = entity.anyDef();
+        auto const *def = entity.anyDef();
         if (!def)
           return;
-        ds->name = def->name(false);
-        ds->detail = def->detailed_name;
-        for (auto &def : entity.def)
-          if (def.file_id == file_id && !ignore(&def)) {
-            kind = ds->kind = def.kind;
-            def_ptrs.push_back(&def);
+        ds.name = def->name(false);
+        ds.detail = def->detailed_name;
+        ds.kind = def->kind;
+
+        if (!ignore(def) && (ds.kind == SymbolKind::Namespace || allows(sym))) {
+          // drop symbols that are behind the current one
+          while (!indent.empty() &&
+                 // use selectionRange for start to start at symbol name
+                 // issue with: int i, j, k; otherwise
+                 indent.top()->range.end < ds.selectionRange.start) {
+            indent.pop();
           }
+          if (indent.empty()) {
+            result.push_back(std::move(ds));
+            indent.push(&result.back());
+          } else {
+            indent.top()->children.push_back(std::move(ds));
+            indent.push(&indent.top()->children.back());
```

Comment:
There is a pointer invalidation issue. I added `    result.reserve(syms.size());`

Reply:
I switched back to `unique_ptr`

Reply:
Oh yeah, thanks, such a common issue, that's why I should have used a vector of pointers.